paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt

```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+            paymentMethod: SupportedPaymentMethod,
+            isSelected: Boolean,
+            isEnabled: Boolean,
+            onItemSelectedListener: () -> Unit
+        ) {
+            composeView.setContent {
+                PaymentMethodUI(
+                    viewWidth = viewWidth,
+                    iconRes = paymentMethod.iconResource,
+                    title = itemView.resources.getString(paymentMethod.displayNameResource),
+                    isSelected = isSelected,
+                    isEnabled = isEnabled,
+                    onItemSelectedListener = onItemSelectedListener
+                )
             }
         }
 
-        fun setEnabled(enabled: Boolean) {
-            binding.root.isEnabled = enabled
-            binding.title.isEnabled = enabled
+        fun onViewRecycled() {
+            // Dispose the underlying Composition of the ComposeView
+            // when RecyclerView has recycled this ViewHolder
+            composeView.disposeComposition()
         }
+    }
+
+    internal companion object {
+        private fun calculateViewWidth(parent: ViewGroup, numberOfPaymentMethods: Int): Dp {
+            val targetWidth = parent.measuredWidth - parent.paddingStart - parent.paddingEnd
+            val screenDensity = parent.context.resources.displayMetrics.density
+            val minItemWidth = 100 * screenDensity + 6.0f + 6.0f
+
+            // if all items fit at min width, then span them across the sheet evenly filling it.
+            // otherwise the number of items visible should be a multiple of .5
+            val viewWidth =
+                if (minItemWidth * numberOfPaymentMethods < targetWidth) {
+                    targetWidth / numberOfPaymentMethods
+                } else {
+                    // numVisibleItems is incremented in steps of 0.5 items
+                    // (1, 1.5, 2, 2.5, 3, ...)
+                    val numVisibleItems = (targetWidth * 2 / minItemWidth).toInt() / 2f
+                    targetWidth / numVisibleItems
+                }
 
-        fun bind(paymentMethod: SupportedPaymentMethod) {
-            binding.icon.setImageResource(paymentMethod.iconResource)
-            binding.title.text = itemView.resources.getString(paymentMethod.displayNameResource)
+            return (viewWidth.toInt() / screenDensity).dp
+        }
+    }
+}
+
+@Composable
+fun PaymentMethodUI(
+    viewWidth: Dp,
+    iconRes: Int,
+    title: String,
+    isSelected: Boolean,
+    isEnabled: Boolean,
+    onItemSelectedListener: () -> Unit
+) {
+    val strokeColor = colorResource(
+        if (isSelected) {
+            R.color.stripe_paymentsheet_add_pm_card_selected_stroke
+        } else {
+            R.color.stripe_paymentsheet_add_pm_card_unselected_stroke
+        }
+    )
+
+    val cardBackgroundColor = if (isEnabled) {
+        MaterialTheme.colors.surface
+    } else {
+        colorResource(R.color.stripe_paymentsheet_elements_background_disabled)
+    }
+
+    Card(
```

Comment:
When we get to adding the Returning user selector, hopefully we can reuse this "Card" component

Reply:
> This is a great first iteration to just refactor not rewrite! I really like the approach here!
> 
> I am curious if you have follow-up PRs related to these changes planned? For instance it seems like the Card element is not reused between the two selectors, and perhaps that is due to current design (it looks like the color selector is different between the two). It seems like this is something that we should resolve, but not necessarily in this review.
> 
> In the past I have tracked things like this in the PR description, JIRA, and a paper doc. I don't have a strong preference, whatever you prefer.

Replying to both. My ideal end goal (will make a tech debt for and circle back after some customization work) is to have a single compose view that replaces the whole recycler view that works for both selectors. We should strive eventually to use the same views for both recyclers. For now I think it's fine to keep them different since their UIs have some nuance that is different.